packages/studio-base/src/components/Chart/datasets.ts

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import * as R from "ramda";
+
+import { TypedData, ObjectData } from "./types";
+
+export type Point = { index: number; x: number; y: number; label: string | undefined };
+
+// Get the length of a typed dataset.
+export function getTypedLength(data: TypedData[]): number {
+  return R.pipe(
+    R.map((v: TypedData) => v.x.length),
+    R.sum,
+  )(data);
+}
+
+/**
+ * iterateObjects iterates over ObjectData, yielding a `Point` for each entry.
+ */
+export function* iterateObjects(dataset: ObjectData): Generator<Point> {
+  let index = 0;
+  for (const datum of dataset) {
+    if (datum == undefined) {
+      index++;
+      continue;
+    }
+
+    const { x, y, label } = datum;
+    yield {
+      index,
+      x,
+      y,
+      label,
+    };
+    index++;
+  }
+}
+
+/**
+ * ExtractPoint maps an object type with array properties to one with the
+ * arrays replaced by their element type. For example:
+ * type Foo = {
+ *   foo: Float32Array;
+ *   bar: number[];
+ *   baz: string[];
+ * }
+ * would be mapped to:
+ * ExtractPoint<Foo> == {
+ *   foo: number;
+ *   bar: number;
+ *   baz: string;
+ * }
+ * It is used to go from `TypedData`'s various incarnations to what a single
+ * point would look like as a `Datum`.
+ *
+ * These `any`s do not introduce anything unsafe; they are necessary for
+ * specifying the type (which is ultimately type-checked at point of use.)
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type ExtractPoint<T extends { [key: string]: Array<any> | Float32Array }> = {
+  [P in keyof T]-?: T[P] extends Float32Array ? number : NonNullable<T[P]>[0];
```

Comment:
I think you can simplify this, because `Float32Array[0]` is `number.

```suggestion
  [P in keyof T]-?: NonNullable<T[P]>[0];
```

Reply:
You're correct! Nice catch!